### ISSUE TITLE
Rules D100, F401, F811, ICN001 cause infinite loop error

### ISSUE THREAD

**Issue by qarmin** [astral-sh/ruff#6786]
Ruff 0.0.285 (latest changes from main branch)

```
ruff  *.py --select ALL --no-cache
```

file content:
```
try:
  import pandas as pa
  import pyarrow as pa
except ImportError:
  pass
```

error:
```
error: Failed to converge after 100 iterations.

This indicates a bug in `ruff`. If you could open an issue at:

    https://github.com/astral-sh/ruff/issues/new?title=%5BInfinite%20loop%5D

...quoting the contents of `PY_FILE_TEST_4291307560.py`, the rule codes D100, F401, F811, ICN001, along with the `pyproject.toml` settings and executed command, we'd be very appreciative!

```

[PY_FILE_TEST_4291307560.py.zip](https://github.com/astral-sh/ruff/files/12411494/PY_FILE_TEST_4291307560.py.zip)

**Comment by zanieb** [comment]
Oscillates between the following

```python
try:
    import pandas as pd
    import pyarrow as pd
except ImportError:
    pass
```
```python
try:
    import pandas as pa
    import pyarrow as pa
except ImportError:
    pass
```

**Comment by zanieb** [comment]
`ICN001` is sufficient to trigger this

**Comment by charliermarsh** [comment]
I think this is actually working "correctly" although we may want to find a way to avoid attempting the fix here. We rename the symbol within the scope, and it's shadowed here. Other editors behave the same way -- if you try to rename the second `pa` here in PyCharm, it will rename both `pa` references:

```python
try:
    import pandas as pa
    import pyarrow as pa
except ImportError:
    pass
```

**Comment by charliermarsh** [comment]
Although that comes with other problems -- for example, this would fail:

```python
if True:
    import pandas as pa
else:
    import pyarrow as pa
```

I _think_ our behavior is correct since I see the same from PyCharm and Pylance (VS Code).

**Comment by qarmin** [comment]
Fix no longer cause problems
